src/extensions/layout-selector/store.js

```diff
@@ -51,9 +77,40 @@ const store = registerStore( 'coblocks/template-selector', {
 	selectors: {
 		isTemplateSelectorActive: ( state ) => state.templateSelector,
 		hasLayouts: ( state ) => !! state.layouts.length,
-		getLayouts: ( state ) => state.layouts || [],
+		getLayouts: ( state ) => {
+			const layouts = state.layouts || [];
+
+			// Taken from Core: https://github.com/WordPress/gutenberg/blob/e41e4f62074fac964d5c92e8836e826e90b289f7/packages/block-editor/src/store/selectors.js#L1434
+			const calculateFrequency = ( time, count ) => {
+				if ( ! time ) {
+					return count;
+				}
+
+				const duration = Date.now() - time;
+
+				switch ( true ) {
+					case duration < MILLISECONDS_PER_HOUR:
+						return count * 4;
+					case duration < MILLISECONDS_PER_DAY:
+						return count * 2;
+					case duration < MILLISECONDS_PER_WEEK:
+						return count / 2;
+					default:
+						return count / 4;
+				}
+			};
+
+			return layouts.map( ( layout ) => {
+				const { time, count = 0 } = state.layoutUsage[ kebabCase( layout.label ) ] || {};
+				return {
+					...layout,
+					frequency: calculateFrequency( time, count ),
```

Comment:
It seems that due to a number of reasons, this function runs a significant number of times within a single load. Since it's using `Date.now()` each subsequent run features a different duration time. If we compare to the Core implementation which features a cached response and thus a `Data.now()` value that can go back in time before the present execution context. 
https://github.com/WordPress/gutenberg/blob/e41e4f62074fac964d5c92e8836e826e90b289f7/packages/block-editor/src/store/selectors.js#L1439-L1440

![calculatedDuration](https://user-images.githubusercontent.com/30462574/103913225-91cf6680-50c5-11eb-9190-1a0cb66a73c9.gif)


Since we are receiving a recent timestamp there are no circumstances where the page execution time should venture beyond `MILLISECONDS_PER_HOUR`. Since we are not accounting for historical usage with this time selector, it makes much of this code unreachable. 

I am of the opinion that this logic should be simplified to apply static incremental updates to the layout selection frequency prop.

Reply:
I agree that this shouldn't be running at this kind of interval. I pulled the `calculateFrequency` function from core to keep the behavior consistent. The PER_HOUR/DAY/WEEK will eventually be hit because the algorithm is based on the current time and the last time a layout was used. So this code is rechable.

I will look into that subsequent run issue.

Reply:
Cool. I see now where the time value is persisted into local storage. 👍

Reply:
After investigating this issue, pulling this function outside of the selector function block and memoizing it seemed to work best. I'm not really sure exactly why it was running so many times but maybe it's related to how the stores work.

Regardless, it only runs once per layout now which is what I expect.

![image](https://user-images.githubusercontent.com/375788/104641897-5e1daf00-5678-11eb-976d-b1b7415b16de.png)